### PR TITLE
Topic Recognition (s47605790): StyleGAN on the ADNI dataset (Task 8)

### PR DESCRIPTION
**Pull request by nguy3ntt**
Hello Teaching Team,

This is my pull request for my project, which involves the implementation of a StyleGAN model on the ADNI dataset using PyTorch (Task 8). The model is designed to generate synthetic brain images similar to those in the ADNI dataset.

Included files and folders:

- dataset.py: Prepares the data for the model
- modules.py: Contains the components of this project
- train.py: Responsible for training the model and generating the loss plot
- predict.py: Used for testing the model and generating the UMAP embedding plot
- README.md: Documentation of the project
- readme_materials: Assets used in the README.md
- All project details are documented in the README.md file.

Thanks,
Tin Trung Nguyen
Student ID: 47605790

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (3 Marks)
Not clear generated images (noise and grids) -1;
Unreasonable losses (even for GAN) -0.5;
Unreasonable embedding (on real/generated images instead of the style code) -0.5

2. Algorithm demonstrated and functions: (2 Marks)
Alaogirhtm demonstrated properly yet does not function as required -1

4. Design: (1 Mark)
Modularized design

5. Commenting: (1 Mark)
Sufficient docstrings and comments

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- The poor generated images suggest the model structure and its training process could be problematic, something I spotted for example:
-- No activation in the mapping network, so no non-linearity to map the `z` to `w`;
-- upsampling in the generator performed without following with a convolution, so the upsampled/stretched values could be in **grid** patterns;
-- training G more often could be helpful to avoid overly powerful D, but it also could introduce noises as it tries to "fool" the D too much without much useful feedback from the D ---- you can try gradient penalty on D instead;
-- scheduler with step 20 could still be too much, you can try "ReduceLROnPlateau" or smaller LRs;

- The UMAP should be plotted on the style code `w`, not the `stacked_images = np.vstack([generated_images_np, np.vstack(all_images)])`. So your UMAP demonstrates difference between real and generated images, not the learned two classes;

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Good introduction of the network. The training loss indicated the model was not yet converged. The generated images are quite noisy and have no clear structures. _The UMAP plot showed a clear separation between generated and real images_, this indicates the inputs of UMAP are generated images and real images, but the input should be style code. UMAP is to estimate AD and NC.
2. **Implementation**: Hyperparameters were hard-coded. Otherwise, it's all good and well-modulized.
3. **Design**: Good
4. **Comment**: Detailed and sufficient
5. **Difficulty**: Hard

**Note**
pycache should not be committed.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Low image quality, unreasonable loss.</td><td>-3</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Unreasonable embedding.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No usage.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. Remove cache files for merge.</td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-6</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by nguy3ntt** [comment]
Hello, thank you for the marking and feedback request. I have made the changes and updated the pull request. Thanks again!

**Comment by shakes76** [comment]
The quality and loss issues -2 +2 for feedback.
